File: core/api/golang/lib/enclaves/partition_connection.go
/*
 * Copyright (c) 2021 - present Kurtosis Technologies Inc.
 * All Rights Reserved.
 */

package enclaves

import (
	"github.com/kurtosis-tech/kurtosis-core/api/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-core/api/golang/lib/binding_constructors"
	"github.com/kurtosis-tech/stacktrace"
)

const (
	unblockedPartitionConnectionPacketLossValue = 0
	blockedPartitionConnectionPacketLossValue   = 100
	//Random packet loss is specified in the 'tc' command in percent. The smallest possible non-zero value is: 2^32 = 0.0000000232% More info: https://wiki.linuxfoundation.org/networking/netem
	smallestPossibleNonZeroPacketLossValue      = 0.0000000232
	maxPossiblePacketLossValue                  = 100
)

// PartitionConnection To get an instance of this type, use the NewUnblockedPartitionConnection, NewBlockedPartitionConnection or NewSoftPartitionConnection functions
type PartitionConnection interface {
	getPartitionConnectionInfo() *kurtosis_core_rpc_api_bindings.PartitionConnectionInfo
}

// ====================================================================================================
//                                    	 Implementations
// ====================================================================================================
// A PartitionConnection implementation
type partitionConnection struct {
	packetLossPercentage float32
}

func (connection *partitionConnection) getPartitionConnectionInfo() *kurtosis_core_rpc_api_bindings.PartitionConnectionInfo {
	partitionConnectionInfo := binding_constructors.NewPartitionConnectionInfo(connection.packetLossPercentage)
	return partitionConnectionInfo
}

// NewUnblockedPartitionConnection returns a PartitionConnection indicating that there is not a network partition
func NewUnblockedPartitionConnection() PartitionConnection {
	return &partitionConnection{
		packetLossPercentage: unblockedPartitionConnectionPacketLossValue,
	}
}

// NewBlockedPartitionConnection returns a PartitionConnection indicating that there is a hard partition
func NewBlockedPartitionConnection() PartitionConnection {
	return &partitionConnection{
		packetLossPercentage: blockedPartitionConnectionPacketLossValue,
	}
}

// NewSoftPartitionConnection returns a PartitionConnection indicating that there is a soft partition with x% percentage of packet loss
func NewSoftPartitionConnection(packetLossPercentage float32) (PartitionConnection, error) {

	if err := isValidPacketLossValue(packetLossPercentage); err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred validating packet loss value")
	}

	partConnection := &partitionConnection{
		packetLossPercentage: packetLossPercentage,
	}

	return partConnection, nil
}

// ====================================================================================================
// 									   Private helper methods
// ====================================================================================================
func isValidPacketLossValue(packetLossPercentage float32) error {

	if packetLossPercentage < smallestPossibleNonZeroPacketLossValue || packetLossPercentage > maxPossiblePacketLossValue {
		return stacktrace.NewError("The packet loss percentage value '%v' is not allowed, it should be >= '%v' and <= '%v'", packetLossPercentage, smallestPossibleNonZeroPacketLossValue, maxPossiblePacketLossValue)
	}

	return nil
}

File: api/typescript/src/core/lib/enclaves/partition_connection.ts
/*
 * Copyright (c) 2021 - present Kurtosis Technologies Inc.
 * All Rights Reserved.
 */

import { PartitionConnectionInfo } from "../../kurtosis_core_rpc_api_bindings/api_container_service_pb";
import { newPartitionConnectionInfo } from "../constructor_calls";

const UNBLOCKED_PARTITION_CONNECTION_PACKET_LOSS_VALUE: number = 0;
const BLOCKED_PARTITION_CONNECTION_PACKET_LOSS_VALUE: number = 100;
//Random packet loss is specified in the 'tc' command in percent. The smallest possible non-zero value is: 2^32 = 0.0000000232% More info: https://wiki.linuxfoundation.org/networking/netem
const SMALLEST_POSSIBLE_NON_ZERO_PACKET_LOSS_VALUE: number = 0.0000000232;
const MAX_POSSIBLE_PACKET_LOSS_VALUE = 100;

// PartitionConnection To get an instance of this type, use the UnblockedPartitionConnection, BlockedPartitionConnection or SoftPartitionConnection objects
export interface PartitionConnection {
    getPartitionConnectionInfo: () => PartitionConnectionInfo;
}

// ====================================================================================================
//                                    	 Implementations
// ====================================================================================================
// UnblockedPartitionConnection use this type of PartitionConnection when you want to establish that the connection is not partitioned
export class UnblockedPartitionConnection {
    private readonly packetLossPercentage: number = UNBLOCKED_PARTITION_CONNECTION_PACKET_LOSS_VALUE;

    constructor(){}

    public getPartitionConnectionInfo(): PartitionConnectionInfo {
        return newPartitionConnectionInfo(this.packetLossPercentage);
    }
}

// BlockedPartitionConnection use this type of PartitionConnection when you want to create a hard partition
export class BlockedPartitionConnection {
    private readonly packetLossPercentage: number = BLOCKED_PARTITION_CONNECTION_PACKET_LOSS_VALUE;

    constructor(){}

    public getPartitionConnectionInfo(): PartitionConnectionInfo {
        let partitionConnectionInfo: PartitionConnectionInfo = new PartitionConnectionInfo();
        partitionConnectionInfo.setPacketLossPercentage(this.packetLossPercentage);
        return partitionConnectionInfo;
    }
}

// SoftPartitionConnection use this type of PartitionConnection when you want to create a partition with x% percentage of packet loss
export class SoftPartitionConnection {
    private readonly packetLossPercentage: number

    constructor(packetLossPercentage: number) {
        if(!SoftPartitionConnection.isValidPacketLossValue(packetLossPercentage)){
            throw new Error(`The packet loss percentage value ${packetLossPercentage} is not allowed, 
            it should be >= ${SMALLEST_POSSIBLE_NON_ZERO_PACKET_LOSS_VALUE} 
            and <= ${MAX_POSSIBLE_PACKET_LOSS_VALUE}`);
        }
        this.packetLossPercentage = packetLossPercentage;
    }

    public getPartitionConnectionInfo(): PartitionConnectionInfo {
        let partitionConnectionInfo: PartitionConnectionInfo = new PartitionConnectionInfo();
        partitionConnectionInfo.setPacketLossPercentage(this.packetLossPercentage);
        return partitionConnectionInfo;
    }

    // ====================================================================================================
    // 									   Private helper methods
    // ====================================================================================================
    private static isValidPacketLossValue(packetLossPercentage: number): boolean {
        return packetLossPercentage >= SMALLEST_POSSIBLE_NON_ZERO_PACKET_LOSS_VALUE && packetLossPercentage <= MAX_POSSIBLE_PACKET_LOSS_VALUE;
    }
}
